core/services/relay/evm/mercury/cache.go

```diff
@@ -0,0 +1,298 @@
+
+type FetchFunc func(ctx context.Context, feedID mercuryutils.FeedID) (*big.Int, error)
+
+type CacheConfig struct {
+	Logger           logger.Logger
+	FetchLatestPrice FetchFunc
+	// LatestPriceTTL controls how "stale" we will allow a price to be e.g. if
+	// set to 1s, a new price will always be fetched if the last result was
+	// from more than 1 second ago.
+	//
+	// Another way of looking at it is such: the cache will _never_ return a
+	// price that was queried from before now-LatestPriceTTL.
+	//
+	// Setting to zero disables caching entirely.
+	LatestPriceTTL time.Duration
+	// MaxStaleAge is that maximum amount of time that a value can be stale
+	// before it is deleted from the cache (a form of garbage collection).
+	//
+	// This should generally be set to something much larger than
+	// LatestPriceTTL. Setting to zero disables garbage collection.
+	MaxStaleAge time.Duration
+}
+
+func NewCache(c CacheConfig) Cache {
+	if c.LatestPriceTTL == 0 {
+		return &NullCache{}
+	}
+	return newMemCache(c)
+}
+
+type cacheVal struct {
+	sync.RWMutex
+
+	fetching bool
+	fetchCh  chan (struct{})
+
+	val *big.Int
+
+	expiresAt time.Time
+}
+
+func (v *cacheVal) read() *big.Int {
+	v.RLock()
+	defer v.RUnlock()
+	return v.val
+}
+
+// memCache stores values in memory
+// it will never return a stale value older than latestPriceTTL, instead
+// waiting for a successful fetch or caller context cancels, whichever comes
+// first
+type memCache struct {
+	services.StateMachine
+	lggr logger.Logger
+
+	latestPriceTTL time.Duration
+	maxStaleAge    time.Duration
+
+	fetchLatestPrice FetchFunc
+
+	cache sync.Map
+
+	wg     sync.WaitGroup
+	chStop utils.StopChan
+}
+
+func newMemCache(c CacheConfig) *memCache {
+	return &memCache{
+		services.StateMachine{},
+		c.Logger.Named("MercuryMemCache"),
+		c.LatestPriceTTL,
+		c.MaxStaleAge,
+		c.FetchLatestPrice,
+		sync.Map{},
+		sync.WaitGroup{},
+		make(chan (struct{})),
+	}
+}
+
+type ErrStopped struct{}
+
+func (e ErrStopped) Error() string {
+	return "memCache was stopped"
+}
+
+// GetLatestPrice
+// FIXME: This will actually block on all types of errors, even non timeouts. Context should be set carefully
+func (m *memCache) GetLatestPrice(ctx context.Context, feedID mercuryutils.FeedID) (*big.Int, error) {
+	vi, _ := m.cache.LoadOrStore(feedID, &cacheVal{
+		sync.RWMutex{},
+		false,
+		nil,
+		nil,
+		time.Now().Add(m.latestPriceTTL),
+	})
+	v := vi.(*cacheVal)
+
+	fWaitForResult := func(ch <-chan struct{}) (*big.Int, error) {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-m.chStop:
+			return nil, ErrStopped{}
+		case <-ch:
+			return v.read(), nil
+		}
+	}
+
+	// HOT PATH
+	v.RLock()
+	if time.Now().After(v.expiresAt) {
+		// if someone else is fetching then wait for the fetch to complete
+		if v.fetching {
+			ch := v.fetchCh
+			v.RUnlock()
+			return fWaitForResult(ch)
+		} else {
+			v.RUnlock()
+			// EXIT HOT PATH
+		}
+	} else {
+		defer v.RUnlock()
+		return v.val, nil
+	}
```

Comment:
Maybe easier to read if you re-order:
```suggestion
	if !time.Now().After(v.expiresAt) {
		defer v.RUnlock()
		return v.val, nil
	} else if v.fetching {
		// if someone else is fetching then wait for the fetch to complete
		ch := v.fetchCh
		v.RUnlock()
		return fWaitForResult(ch)
	}
	v.RUnlock()
```